packages/jest-mock/src/__tests__/index.test.ts

```diff
@@ -1308,6 +1308,15 @@ describe('moduleMocker', () => {
       expect(() => {
         moduleMocker.spyOn(null, 'method');
       }).toThrow('spyOn could not find an object to spy on for method');
+      expect(() => {
+        moduleMocker.spyOn({} as Record<string, any>, '');
+      }).toThrow('No property name supplied');
+      expect(() => {
+        moduleMocker.spyOn({} as Record<number, any>, NaN);
+      }).toThrow('No property name supplied');
+      expect(() => {
+        moduleMocker.spyOn({}, undefined);
+      }).toThrow('No property name supplied');
       expect(() => {
         moduleMocker.spyOn({}, 'method');
       }).toThrow(
```

Comment:
In a way these message (and also some of "Cannot spy on the method property because it is not a function" below) are somewhat misleading. How about reworking the whole check logic like this or so:

```ts
if (methodKey == null) {
  throw new Error('No property name supplied');
}

if (accessType) {
  return this._spyOnProperty(object, methodKey, accessType);
}

const original = object[methodKey];

if (!original) {
  throw new Error(
    `Property ${String(methodKey)} does not exist in the provided object`,
  );
}
```

Reply:
E.g., `spyOn` getters or setters has property existence check:

https://github.com/facebook/jest/blob/15af9b3b6f53a4f2f2e39d328eb6ef9f46ba91c4/packages/jest-mock/src/index.ts#L1268-L1270

Reply:
If you ask me, I have no objections. But the messages you refer to were implemented back in 2018 (#5107) for reasons I'm not aware of, and changing those go beyond the scope of fixing the related bug and this PR, I believe. Feel free to do it in a separate PR though, but please note that the code you suggested will _not_ fix the bug and will actually bring it back, because `methodKey==null` is an incorrect way to check for an object key, since it's `true` for `methodKey===0`)

Reply:
> since it's `true` for `methodKey===0`

Not sure I follow this. `methodKey == null` would exclude only `undefined` and `null`, hence the message `'No property name supplied'` is correct. This message does not sound correct if user passed `''` or `NaN` as a `methodKey`. Do I miss something?

Reply:
> Feel free to do it in a separate PR though

See #14087